internal/pkg/skuba/addons/metricsserver.go

```diff
@@ -0,0 +1,297 @@
+/*
+ * Copyright (c) 2020 SUSE LLC.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package addons
+
+import (
+	"bufio"
+	"encoding/base64"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/pkg/errors"
+	kubeadmconstants "k8s.io/kubernetes/cmd/kubeadm/app/constants"
+	"k8s.io/kubernetes/cmd/kubeadm/app/images"
+
+	"github.com/SUSE/skuba/internal/pkg/skuba/kubernetes"
+	"github.com/SUSE/skuba/internal/pkg/skuba/metricsserver"
+	"github.com/SUSE/skuba/internal/pkg/skuba/skuba"
+	skubaconstants "github.com/SUSE/skuba/pkg/skuba"
+)
+
+func init() {
+	registerAddon(kubernetes.MetricsServer, renderMetricsServerTemplate, metricsServerCallbacks{}, normalPriority, []getImageCallback{GetMetricsServerImage})
+}
+
+func GetMetricsServerImage(imageTag string) string {
+	return images.GetGenericImage(skubaconstants.ImageRepository, "metrics-server", imageTag)
+}
+
+func (renderContext renderContext) MetricsServerImage() string {
+	return GetMetricsServerImage(kubernetes.AddonVersionForClusterVersion(kubernetes.MetricsServer, renderContext.config.ClusterVersion).Version)
+}
+
+func (renderContext renderContext) CABundle() string {
+	// base64 encoded Kubernetes CA certificate
+	path := filepath.Join(skubaconstants.PkiDir(), kubeadmconstants.CACertName)
+	fi, err := os.Stat(path)
+	if os.IsNotExist(err) {
+		return ""
+	}
+	if fi.IsDir() {
+		return ""
+	}
+
+	f, err := os.Open(path)
+	if err != nil {
+		return ""
+	}
+
+	reader := bufio.NewReader(f)
+	content, err := ioutil.ReadAll(reader)
+	if err != nil {
+		return ""
+	}
+
+	return base64.StdEncoding.EncodeToString(content)
+}
```

Comment:
The `caBundle` renders the Kubernetes CA with base64 encoded.

When `skuba cluster init`, at this time Kubernetes CA has not generated yet, render the `caBundle` empty string.

When `skuba bootstrap` and `skuba addon upgrade apply`, the Kubernetes CA file had generated. When apply addon, the manifest would be re-render again. So the `caBundle` rendered with base64 Kubernetes CA.

Reply:
@jenting this information is useful. Suggest to add as function comment.

Reply:
@jenting this information is useful. Suggest to add as function comment.